sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/ShardSubscribersPoolImpl.java

```diff
@@ -0,0 +1,404 @@
+  private static final long BUFFER_POLL_WAIT_MS = 1_000L;
+  private static final long ENQUEUE_TIMEOUT_MS = 3_000L;
+
+  private final UUID poolId;
+  private final Config config;
+  private final AsyncClientProxy kinesis;
+  private final KinesisReaderCheckpoint initialCheckpoint;
+  private final ConcurrentMap<String, ShardSubscriberState> shardsStates;
+  private final BlockingQueue<ExtendedKinesisRecord> eventsBuffer;
+
+  public ShardSubscribersPoolImpl(
+      Config config, AsyncClientProxy kinesis, KinesisReaderCheckpoint initialCheckpoint) {
+    this.poolId = UUID.randomUUID();
+    this.config = config;
+    this.kinesis = kinesis;
+    this.initialCheckpoint = initialCheckpoint;
+    this.shardsStates = new ConcurrentHashMap<>();
+    this.eventsBuffer = new LinkedBlockingQueue<>(BUFFER_SIZE);
+  }
+
+  @Override
+  public UUID getPoolId() {
+    return poolId;
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public boolean start() {
+    // TODO: handle different types of start:
+    // ShardIteratorType.LATEST
+    // ShardIteratorType.AT_SEQUENCE_NUMBER
+    // ShardIteratorType.TRIM_HORIZON
+    // ShardIteratorType.AT_TIMESTAMP
+    // AT_SEQUENCE_NUMBER:
+    // - check if we have seen this seq number before
+    // - check if we have seen a sub-sequence number before
+
+    List<String> startingWithShards = checkpointToShardsIds(initialCheckpoint);
+    LOG.info(
+        "Starting pool {} {} {}. Shards = {}",
+        poolId,
+        config.getStreamName(),
+        config.getConsumerArn(),
+        startingWithShards);
+
+    initialCheckpoint.forEach(
+        shardCheckpoint -> createSubscription(shardCheckpoint.getShardId(), shardCheckpoint, true));
+    return true;
+  }
+
+  private void decommissionShardSubscription(String shardId) {
+    shardsStates.computeIfPresent(
+        shardId,
+        (s, st) -> {
+          st.cancel();
+          st.markClosed();
+          return st;
+        });
+  }
+
+  @Override
+  public boolean stop() {
+    KinesisReaderCheckpoint ch = getCheckpointMark();
+    List<String> stoppingWithShards = checkpointToShardsIds(ch);
+    LOG.info(
+        "Stopping pool {} {} {}. Shards = {}",
+        poolId,
+        config.getStreamName(),
+        config.getConsumerArn(),
+        stoppingWithShards);
+
+    shardsStates.values().forEach(ShardSubscriberState::cancel);
+    try {
+      Thread.sleep(config.getStopCoolDownDelayMs());
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+    return true;
+  }
+
+  /**
+   * This is ultimately called by Netty threads via {@link KinesisShardEventsSubscriber} callbacks.
+   *
+   * @param event
+   * @throws InterruptedException
+   */
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void handleEvent(ShardEvent event) throws InterruptedException {
+    switch (event.getType()) {
+      case SUBSCRIPTION_COMPLETE:
+        CompletableFuture.runAsync(
```

Comment:
Wrapping everything into `CompletableFuture.runAsync` isn't really what we want. Think of it this way: 

Instead of us waiting / blocking until a call to Kinesis returns or we've received an event to do something, we want to be called at that time and react accordingly.

The way to to this is to register a completion callback handler on the `CompletableFuture`. Below an example of an async re-subscription loop. The `CompletableFuture` returned by `subscribe` will only ever be completed normally if the shard is completely consumed or completed exceptionally if the subscription fails.

```Java
final CompletableFuture<Void> completed = new CompletableFuture<>();

CompletableFuture<Void> subscribe(StartingPosition position) {
  try {
    kinesis.subscribeToShard(subscribeRequest(position), responseHandler())
        .whenCompleteAsync(
            (Void unused, Throwable error) -> {
              if(error!= null){
                // FIXME potentially resubscribe
                completed.completeExceptionally(error);
              } else {
                String lastContinuationSequenceNumber = ...
                if(lastContinuationSequenceNumber == null){
                  completed.complete(null); // completely consumed this shard, done
                } else {
                  subscribe(StartingPosition.builder()
                      .type(ShardIteratorType.AT_SEQUENCE_NUMBER)
                      .sequenceNumber(lastContinuationSequenceNumber)
                      .build());
                }
              }
            });
    return completed;
  } catch (RuntimeException e) {
    completed.completeExceptionally(e);
    return completed;
  }
}


```

Reply:
We could also use `whenComplete` in this case as our callback code is cheap and never blocking. The difference to `whenCompleteAsync` is the thread our callback code will be running on:
- `whenComplete` runs our code on the calling thread, likely a Netty thread in the case above.
- `whenCompleteAsync` runs our code on a new thread in the common ForkJoinPool.